src/main/java/com/internhub/data/positions/scrapers/impl/DepthAwarePositionScraper.java

```diff
@@ -0,0 +1,133 @@
+package com.internhub.data.positions.scrapers.impl;
+
+import com.google.common.collect.Lists;
+import com.internhub.data.models.Company;
+import com.internhub.data.models.Position;
+import com.internhub.data.positions.pages.Page;
+import com.internhub.data.positions.pages.PageScraper;
+import com.internhub.data.positions.scrapers.QueuedPositionScraper;
+import com.internhub.data.positions.scrapers.candidates.CandidateComparator;
+import com.internhub.data.positions.scrapers.candidates.DepthAwareCandidate;
+import com.internhub.data.util.ScraperUtils;
+import com.internhub.data.search.GoogleSearch;
+import org.openqa.selenium.TimeoutException;
+import org.openqa.selenium.WebDriver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URL;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class DepthAwarePositionScraper extends QueuedPositionScraper<DepthAwareCandidate> {
+    private static final String INTERNSHIP_SEARCH_TERM = "%s internship apply";
+
+    private static final int MAX_DEPTH = 4;
+    private static final int TOP_N_GOOGLE_LINKS = 4;
+    private static final int MAX_TOTAL_LINKS = 100;
+    private static final int PAGE_LOAD_DELAY = 2000;
+
+    private static final Logger logger = LoggerFactory.getLogger(DepthAwarePositionScraper.class);
+
+    private WebDriver mDriver;
+    private GoogleSearch mGoogle;
+    private int mTotalLinks;
+    private PageScraper mPageScraper;
+    private PageScraper.PageScrapeResult mLastPageResult;
```

Comment:
I see what you are trying to do: you are trying to abstract away the entire process of searching from these scrapers. I don't agree with that - the code becomes pretty messy. Look at what you are doing with the `mLastPageResult` field - you have to set it each time you scrape a page just so you can return the next candidates. It's very clunky and the reason it's this way is because you are tightly bound by the inheritance contract. My suggestion is just use an interface for these scrapers. The interfaces before were fine, you should let these classes have their own implementation instead of abstracting the logic away.

Also if there were more than one implementation of a QueuedPositionScraper (or even plans to) then I would be willing to change my mind. But how would that work? Say you wanted a QueuedPositionScraper that kept track of link degree or some other field so you made `LinkDegreePositionScraper` but you also wanted it to keep track of depth then you would have `LinkDegreeDepthAwarePositionScraper`? Inheritance like this is very restrictive and it will lead to our code ballooning when we want to add more features.